aten/src/ATen/native/TensorIterator.cpp

```diff
@@ -147,23 +181,40 @@ void TensorIterator::compute_types() {
           op.dtype = common_dtype;
         }
       }
-    }
-  }
 
-  for (auto& op : operands_) {
-    auto& tensor = op.tensor;
-    if (!tensor.defined()) {
-      continue;
-    }
-    if (op.device != tensor.device() || op.dtype != tensor.scalar_type()) {
-      if (op.is_output) {
-        AT_ERROR("output with device ", tensor.device(), " and dtype ", tensor.scalar_type(),
-                 " doesn't match the desired device ", op.device, " and dtype ", op.dtype);
-      } else if (tensor.dim() == 0) {
-        tensor = tensor.to(op.options());
-      } else {
-        AT_ERROR("expected device ", op.device, " and dtype ", op.dtype,
-                 " but got device ", tensor.device(), " and dtype ", tensor.scalar_type());
+      if (op.tensor.defined() && op.tensor.scalar_type() != common_dtype)
+      {
+        if (op.is_output) {
+          if (!can_promote(common_dtype, op.tensor.scalar_type())) {
+            AT_ERROR("result type ", common_dtype,
+              " doesn't match the desired output type ",
+              op.tensor.scalar_type());
+          }
+        } else {
+          auto tensor = op.tensor;
+          op.tensor = tensor.to(common_dtype);
+          op.dtype = op.tensor.scalar_type();
+          auto original_element_size = tensor.element_size();
+          auto new_element_size = op.tensor.element_size();
+          // stride size (in bytes) can change if we change the dtype.
+          if ( original_element_size != new_element_size ) {
```

Comment:
In general, if you make a copy of a tensor (via tensor.to()) you need to recalculate the strides even if the element size doesn't change.

Reply:
ok good to know